### PR TITLE
Compile-time reduction - file request variant

### PR DESCRIPTION
**Pull request by nemerle**
Do not instantiate std::variant - FileRequest.h usages moved to cpp files

### CODE REVIEW

**Review by lawsonamzn — APPROVED** [review]
Looks reasonable to me in terms of cutting down on compile time.  Noticed there were a bunch of whitespace-only changes though, perhaps a setting on your text editor is cutting out spaces and inserting them?

File: Code/Framework/AzCore/AzCore/IO/Streamer/BlockCache.h [hunk 1]
```diff
@@ -23,6 +23,11 @@ namespace AZ
 {
     namespace IO
     {
+        class RequestPath;
+        namespace Requests {
+            struct ReadData;
+        }
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/IO/Streamer/BlockCache.h:29]: coding style: The opening curly brace for the namespace should be on it's own line

File: Code/Framework/AzCore/AzCore/IO/Streamer/FileRequest.cpp [hunk 2]
```diff
@@ -12,117 +12,124 @@
...
-        , m_priority(priority)
-        , m_memoryType(IStreamerTypes::MemoryType::ReadWrite) // Only generic memory can be assigned externally.
-    {}
-
-    FileRequest::ReadRequestData::~ReadRequestData()
-    {
-        if (m_allocator != nullptr)
+namespace AZ::IO::Requests
+{
+ReadData::ReadData(void* output, u64 outputSize, const RequestPath& path, u64 offset, u64 size, bool sharedRead)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/IO/Streamer/FileRequest.cpp:21]: This namespace is missing an indent on it's contents

File: Code/Framework/AzCore/AzCore/IO/Streamer/DedicatedCache.h [hunk 3]
```diff
@@ -11,84 +11,88 @@
...
-            AZStd::vector<FileRange> m_cachedFileRanges;
-            AZStd::vector<AZStd::unique_ptr<BlockCache>> m_cachedFileCaches;
-            AZStd::vector<size_t> m_cachedFileRefCounts;
-
-            AZ::Statistics::RunningStatistic m_usagePercentageStat;
+        struct CreateDedicatedCacheData;
+        struct DestroyDedicatedCacheData;
+    } // namespace Requests
+
+    struct DedicatedCacheConfig final : public IStreamerStackConfig
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/IO/Streamer/DedicatedCache.h:29]: It is preferred to follow the "chameleon rule" when making changes regard the style of a file to avoid making unnecessary changes.
So the base class moving from a separate line to a line on it's own seems kind of strange.

File: Code/Framework/AzCore/AzCore/IO/Streamer/FileRequest.h [hunk 4]
```diff
@@ -21,403 +21,417 @@
...
-        class StreamStackEntry;
-        class ExternalFileRequest;
+    class StreamStackEntry;
+    class ExternalFileRequest;
 
-        using FileRequestPtr = AZStd::intrusive_ptr<ExternalFileRequest>;
-        
-        class FileRequest final
+    using FileRequestPtr = AZStd::intrusive_ptr<ExternalFileRequest>;
+    namespace Requests
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/IO/Streamer/FileRequest.h:30]: Alternatively for nesting inner namespace blocks is to open a namespace block just for the forward declarations and type aliases and then use the nested namespace feature to open the inner namespace blocks with the same indentation.

```
namespace AZ::IO
{
    class StreamStackEntry;
    class ExternalFileRequest;

    using FileRequestPtr = AZStd::intrusive_ptr<ExternalFileRequest>;
}

namespace AZ::IO::Requests
{
    // ...
}

File: Code/Framework/AzCore/AzCore/IO/Streamer/Scheduler.h [hunk 5]
```diff
@@ -8,6 +8,7 @@
 
 #pragma once
 
+#include <AzCore/IO/Streamer/FileRequest.h>
 #include <AzCore/IO/IStreamerTypes.h>
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/IO/Streamer/Scheduler.h:12]: It looks like the need for this include is only for accessing the `FileRequestPtr` type alias.
In theory this file could make another type alias for it to avoid the forward declaration
i.e
```
namespace AZ::IO
{
    class ExternalFileRequest;
    inline namespace SchedularInternal
    {
        using ExternalRequestPtr = AZStd::intrusive_ptr<ExternalFileRequest>;
    }
}
```

File: Code/Framework/AzCore/AzCore/IO/Streamer/StreamerContext.h [hunk 6]
```diff
@@ -7,126 +7,127 @@
...
+#include <AzCore/std/containers/vector.h>
 
-namespace AZ
+namespace AZ::IO
 {
-    namespace IO
+    class FileRequest;
+    class ExternalFileRequest;
+
+    using FileRequestPtr = AZStd::intrusive_ptr<ExternalFileRequest>;
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/IO/Streamer/StreamerContext.h:24]: Well this will cause a problem if both `FileRequests.h` and `StreamerContext.h` was included in the same file at any point.

If anything you could use an `inline namespace` in both files to distinguish the file aliases.
That way if there is ever a situation where both files need to be included into the same location, the developer can resolve the `FileRequestPtr` by explicitly indicating the namespace

Ex. File with either StreamerContext.h or FileRequest.h `include`ed
```
void Foo(AZ::IO::FileRequestPtr);
```

Ex. File with both StreamerContext.h and FileRequest.h `include`ed

```
// Pick one of the type aliases to resolve ambiguity
void Foo(AZ::IO::StreamerContextInternal::FileRequestPtr);
```

File: Code/Framework/AzCore/Platform/Windows/AzCore/IO/Streamer/StorageDrive_Windows.cpp [hunk 7]
```diff
@@ -257,13 +257,13 @@ namespace AZ::IO
             hasWorked = AZStd::visit([this, request](auto&& args)
             {
                 using Command = AZStd::decay_t<decltype(args)>;
-                if constexpr (AZStd::is_same_v<Command, FileRequest::FileExistsCheckData>)
+                    if constexpr (AZStd::is_same_v<Command, Requests::FileExistsCheckData>)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Platform/Windows/AzCore/IO/Streamer/StorageDrive_Windows.cpp:260]: There is an extra indent here

File: Code/Framework/AzCore/AzCore/IO/Streamer/StreamerContext.h (hunk 6 above)
Comment (nemerle) [reply Code/Framework/AzCore/AzCore/IO/Streamer/StreamerContext.h:24]: I don't believe it will cause problems, since multiple `using` with the same type/name are allowed?
<https://godbolt.org/z/Ee8eofzYx>

**Comment by nemerle** [comment]
Following the suggestions I've stumbled on additional reduction that is related to the original change :

* Reduce the number of includes in AssetDataStream.h
* PIMPL away members of AssetDataStream that needed those heavy include files.
* Move ReportType out of ReportData to allow forward declarations.
* Move Scheduler destructor to cpp file -> reduce the need for full class definitions

File: Code/Framework/AzCore/AzCore/IO/Streamer/DedicatedCache.h (hunk 3 above)
Comment (nemerle) [reply Code/Framework/AzCore/AzCore/IO/Streamer/DedicatedCache.h]: I understand the rule, the clang-format is fighting me though :smile: 

Maybe we should consider running clang-format on all the source files to prevent such things?

File: Code/Framework/AzCore/AzCore/IO/Streamer/BlockCache.h (hunk 1 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/IO/Streamer/BlockCache.h]: +1

File: Code/Framework/AzCore/AzCore/IO/Streamer/FileRequest.cpp (hunk 2 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/IO/Streamer/FileRequest.cpp]: +1 (using a selection with `Ctrl-K, F` in Visual Studio will run clang-format to fix this too 🙂)

File: Code/Framework/AzCore/AzCore/IO/Streamer/DedicatedCache.h (hunk 3 above)
Comment (hultonha) [reply Code/Framework/AzCore/AzCore/IO/Streamer/DedicatedCache.h]: Yeah we totally should, some files have it and some do not, for those that do not the best way is to just select the updated lines and use `Ctrl-K, F`. It might be worth creating a format-only PR (for a small number of files at a time) to update things to stop functional and form changes getting mixed in together

**Review by hultonha — COMMENTED** [review]
I started having a look at the PR but it's very very large... is there any way to split it up to make reviewing a little more reasonable? There's also what seems to be formatting changes mixed in with functional changes which can make it harder to know what to focus on, if we can in future try to keep those to a minimum that would be super helpful.

**Comment by nemerle** [comment]
@hultonha I'll try to split this up into 2 PRs, 
About the whitespace changes, some of them were caused by namespace grouping, some by using clang-format on files that are non-conforming with our style and some by IDE removing spurious spaces, how should I handle that correctly?

File: Code/Framework/AzCore/AzCore/IO/Streamer/StreamerContext.h (hunk 6 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/IO/Streamer/StreamerContext.h:24]: You are [correct](https://en.cppreference.com/w/cpp/language/typedef) about a type alias being able to re-declared to the same type.
> The typedef-names are aliases for existing types, and are not declarations of new types. Typedef cannot be used to change the meaning of an existing type name (including a typedef-name). Once declared, a typedef-name may only be redeclared to refer to the same type again

**Comment by amznestebanpapp** [comment]
> > @hultonha I'll try to split this up into 2 PRs, About the whitespace changes, some of them were caused by namespace grouping, some by using clang-format on files that are non-conforming with our style and some by IDE removing spurious spaces, how should I handle that correctly?
> 
> Thank you @nemerle, that would be super helpful.
> 
> So the namespace grouping ones...

Also note that contributors are currently not "forced" to run clang-format. So formatting issues will keep going in.
Maybe something to bring up to the sigs... if we want to reduce the amount of issues we see in PRs regarding formatting (which I would agree on) we can do a pass (as @hultonha is suggesting) and also add a validation rule that checks that formatting has been applied to new changes (this is the validation job that gets run to approve a PR)

**Review by amznestebanpapp — APPROVED** [review]
Do you have some numbers on how much this improves compilation times? (a before/after of a full profile build is enough for me)

**Comment by nemerle** [comment]
> Do you have some numbers on how much this improves compilation times? (a before/after of a full profile build is enough for me)

Non-unity build ~2% faster compilation, ~0.5% smaller binary folder ( less template debug info )

**Comment by nemerle** [comment]
Quick question to my kind reviewers, do You perhaps take advantage of Github's 'hide whitespace changes' feature during the review?

**Comment by nemerle** [comment]
Closing this PR as it will be split into two: #6192 for most of the whitespace changes, and another for functional ones.

**Comment by hultonha** [comment]
> Quick question to my kind reviewers, do You perhaps take advantage of Github's 'hide whitespace changes' feature during the review?

That's a great idea, good reminder! I'll do that for the formatting changes to help avoid some of the noise. Thanks

> Closing this PR as it will be split into two: #6192 for most of the whitespace changes, and another for functional ones.

Amazing thank you @nemerle, sorry for the hassle but really appreciate the effort. Please link the new PRs here when they're ready and I'll take a look ASAP 👍

**Comment by nemerle** [comment]
The split PRs are available now ( links in my previous comment message).

Please note, the second PR is based on the first PRs branch, so before the first one is merged, the second one will look a bit 'fluffier' than it is in reality :smile:

**Comment by hultonha** [comment]
Awesome! 😄 Thanks @nemerle! I'll take a look shortly 🙂 Haha got you, yeah that branch can be rebased onto development again after the formatting changes go in (or however you'd like to do it). Cheers!

**Comment by amznestebanpapp** [comment]
Another option for future PRs is to put the formatting changes in a different commit, then the reviewers can filter the changes that way. 
I really don't want PR authors or reviewers wasting time on formatting.
I will talk with both of you on discord to see if we can just do a formatting pass and add something into the PR pipeline that formats new changes the expected way.
